contracts/optimizedForceMove.sol

```diff
@@ -0,0 +1,234 @@
+            )
+        );
+        address challenger = _recoverSigner(
+            msgHash,
+            challengerSig.v,
+            challengerSig.r,
+            challengerSig.s
+        );
+        require(
+            _isAddressInArray(challenger, fixedPart.participants),
+            'Challenger is not a participant'
+        );
+
+        // ------------
+        // EFFECTS
+        // ------------
+
+        channelStorage = ChannelStorage(
+            largestTurnNum,
+            now + fixedPart.challengeDuration,
+            stateHashes[m],
+            challenger,
+            keccak256(abi.encode(variableParts[m].outcome))
+        );
+
+        channelStorageHashes[channelId] = keccak256(abi.encode(channelStorage));
+    }
+    // Internal methods:
+
+    function _isAddressInArray(address suspect, address[] memory addresses)
+        internal
+        pure
+        returns (bool)
+    {
+        for (uint256 i = 0; i < addresses.length; i++) {
+            if (suspect == addresses[i]) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    function _validSignatures(
+        uint256 largestTurnNum,
+        address[] memory participants,
+        bytes32[] memory stateHashes,
+        Signature[] memory sigs,
+        uint8[] memory whoSignedWhat // whoSignedWhat[i] is the index of the state in stateHashes that was signed by participants[i]
+    ) internal pure returns (bool) {
+        uint256 nParticipants = participants.length;
+        uint256 nStates = stateHashes.length;
+
+        require(
+            whoSignedWhat.length == nParticipants,
+            '_validSignatures: whoSignedWhat must be the same length as participants'
+        );
+
+        require(
+            _acceptableWhoSignedWhat(whoSignedWhat, largestTurnNum, nParticipants, nStates),
+            'Unacceptable whoSignedWhat array'
+        );
+        for (uint256 i = 0; i < nParticipants; i++) {
+            address signer = _recoverSigner(
+                stateHashes[whoSignedWhat[i]],
+                sigs[i].v,
+                sigs[i].r,
+                sigs[i].s
+            );
+            if (signer != participants[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    function _acceptableWhoSignedWhat(
+        uint8[] memory whoSignedWhat,
+        uint256 largestTurnNum,
+        uint256 nParticipants,
+        uint256 nStates
+    ) internal pure returns (bool) {
+        for (uint256 i = 0; i < nParticipants; i++) {
+            uint256 offset = (nParticipants + largestTurnNum - i) % nParticipants;
+            // offset is the difference between the index of the current participant and the index of the participant who owns the largesTurnNum state
+            // the extra nParticipants ensures offset always positive
+            if (offset < nStates) {
+                if (whoSignedWhat[i] < nStates - offset) {
```

Comment:
You could probably combine these by writing `offset + whoSignedWhat[i] < nStates`.

Reply:
👍